arrow-fx/src/test/kotlin/arrow/fx/QueueTest.kt

```diff
@@ -107,7 +171,72 @@ class QueueTest : UnitSpec() {
             !q.offer(i)
             !q.shutdown()
             !q.take()
-          }.attempt().unsafeRunSync() == Left(QueueShutdown)
+          }.equalUnderTheLaw(IO.raiseError(QueueShutdown), EQ())
+        }
+      }
+
+      "$label - time out peeking from an empty queue" {
+        IO.fx {
+          val wontComplete = queue(10).flatMap(Queue<ForIO, Int>::peek)
+          val start = !effect { System.currentTimeMillis() }
+          val received = !wontComplete.map { Some(it) }
+            .waitFor(100.milliseconds, default = just(None))
+          val elapsed = !effect { System.currentTimeMillis() - start }
+          !effect { received shouldBe None }
+          !effect { (elapsed >= 100) shouldBe true }
+        }.equalUnderTheLaw(IO.unit, EQ())
+      }
+
+      "$label - suspended peek calls on an empty queue complete when offer calls made to queue" {
+        forAll(Gen.int()) { i ->
+          IO.fx {
+            val q = !queue(3)
+            val first = !q.peek().fork(ctx)
+            !q.offer(i)
+            val res = !first.join()
+            !effect { res shouldBe i }
+          }.equalUnderTheLaw(IO.unit, EQ())
+        }
+      }
+
+      "$label - multiple peek calls on an empty queue all complete with the first value is received" {
+        forAll(Gen.int()) { i ->
+          IO.fx {
+            val q = !queue(1)
+            val first = !q.peek().fork(ctx)
+            val second = !q.peek().fork(ctx)
+            val third = !q.peek().fork(ctx)
+            !q.offer(i)
+            val firstValue = !first.join()
+            val secondValue = !second.join()
+            val thirdValue = !third.join()
+            !effect { setOf(firstValue, secondValue, thirdValue) shouldBe setOf(i, i, i) }
+          }.equalUnderTheLaw(IO.unit, EQ())
+        }
+      }
+
+      "$label - peeking a shutdown queue creates a QueueShutdown error" {
+        forAll(Gen.int()) { i ->
+          IO.fx {
+            val q = !queue(10)
+            !q.offer(i)
+            !q.shutdown()
+            !q.peek()
+          }.equalUnderTheLaw(IO.raiseError(QueueShutdown), EQ())
+        }
+      }
+
+      "$label - peek does not remove value from Queue" {
+        forAll(Gen.int()) { i ->
+          IO.fx {
+            val q = !queue(10)
+            val size1 = !q.size()
+            !q.offer(i)
+            val size2 = !q.size()
+            val peeked = !q.peek()
+            val size3 = !q.size()
+            !effect { Tuple4(size1, size2, peeked, size3) shouldBe Tuple4(0, 1, i, 1) }
+          }.equalUnderTheLaw(IO.unit, EQ())
```

Comment:
Yes of course, but after the last size check the element in the queue may or may not be the one peek returned. Logically it obviously is, it is better to have that tested though. Instead of checking the size after `peek`, why not use `takeAll` and verify that it is equal to `listOf(i)`?

Reply:
Good point, update the test. To first `offer`, `peek` and `takeAll` and checks the product of `peek` and `takeAll`.